### PR TITLE
Major refactor (API + SDK)

### PR DESCRIPTION
**Pull request by claudiolemos**
This pull request will include major refactor of the API + SDK, so I'll need everyone's help to discuss better ways to improve the following commits.

Major changes that you can test as of now.

There's now an .env file where you'll store major api keys as well as some configs for servers. You can edit the `.env.template` and remove the template extension

```python
OPENAI_API_KEY="your-openai-api-key"
ANTHROPIC_API_KEY="your-anthropic-api-key"
COHERE_API_KEY="your-cohere-api-key"
ENGINE_HOST="localhost"
ENGINE_PORT=8000
UI_HOST = "localhost"
UI_PORT = 3000
LOG_LEVEL = "info"
```

Afterwards you can test the following new API routes on Postman:

1. **`GET` http://localhost:8000/health**: health check

2. **`GET` http://localhost:8000/api/engine/models**: list all models (you can add param `provider` (e.g.: `openai`), to filter

3. **`GET` http://localhost:8000/api/engine/providers**: list all providers
4. **`POST`http://localhost:8000/api/engine/chat/{provider}**: chat with provide (`openai`,`cohere`,`anthropic`)
   ```json
   {
      "model": "gpt-4",
      "chat_input": "Say hi back"
      // "api_key": "your-api-key" # not necessary if you set .env file
      // "stream": true # default is false
   }
   ```

You can also run the SDK (only OpenAI and Anthropic, for now)

```python
anthropic_client = AnthropicClient(api_key="anthropic_api_key")
openai_client = OpenAIClient(api_key="openai_api_key")
claude2 = anthropic_client.get_model(model="claude-2")
gpt3 = openai_client.get_model(model="gpt-3.5-turbo")
gpt4 = openai_client.get_model(model="gpt-4")

gpt3.chat("Say hi back")

response = gpt4.chat("Hi write a paragfraph about space", is_stream=True)
for chunk in response:
    print(chunk, end="", flush=True)
```

So to test this refactor do the following:

1. set your env file
2. run llmstudio server
3. run all requests for all provider (non-stream and streamed) on POSTMAN
4. run the notebook

### CODE REVIEW

**Comment by claudiolemos** [comment]
I want to open discussion mostly with @reiid00 @MiNeves00 @gabrielrfg regarding syntax for the SDK

- should we have a unique client for each provider? if so, what's the benefit of that?
- or could we just have a unique chat function? `llmstudio.chat(model="openai/gpt-3.5-turbo", input="Tell me a joke")`

**Comment by claudiolemos** [comment]
I removed (for now!) the `_select_appropriate_model` and `max_retries` from OpenAI + VertexAI, but both will be implemented back. We should consider extending max_retries to other providers.

**Comment by MiNeves00** [comment]
The issue with doing this: `llmstudio.chat(model="openai/gpt-3.5-turbo", input="Tell me a joke")` is that for each chat you would need to repeat the model

So if we were to go down this route I would maybe say to just join the client and the .get_model together like this: `model=LLM(model="openai/gpt3.5-turbo")` but still do `model.chat("Hi")`

**Review by gabrielrfg — APPROVED** [review]

**Review by MiNeves00 — APPROVED** [review]
Some functionalities were temporarily removed that should be readded in the future, such as:
- server creation in sdk
- Langchain LLM Wrapper

File: llmstudio/engine/__init__.py [hunk 1]
```diff
@@ -176,24 +78,69 @@ def create_app_from_config(config: EngineRouteConfig) -> EngineAPI:
...
+            for provider_config in config.providers.values():
+                if provider_config.models:
+                    all_models.extend(provider_config.models.keys())
+            return list(set(all_models))  # Use set to avoid duplicates if any
+
+    # Function to create a chat handler for a provider
+    def create_chat_handler(provider_config):
+        async def chat_handler(request: Request):
+            """Endpoint for chat functionality."""
+            provider_class = globals()[f"{provider_config.name}Provider"]
```
[the remark this answers is not in the document]
Comment (claudiolemos) [reply llmstudio/engine/__init__.py]: Added @provider decorator + provider registry

**Pull request merged**